File: apina-java/src/main/java/fi/evident/apina/java/reader/ClassMetadataReader.java
package fi.evident.apina.java.reader;

import fi.evident.apina.java.model.*;
import fi.evident.apina.java.model.type.JavaBasicType;
import org.objectweb.asm.*;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

import static java.util.Objects.requireNonNull;
import static java.util.stream.Collectors.toList;

/**
 * Reads class metadata.
 */
final class ClassMetadataReader {

    public static JavaClass loadMetadata(InputStream in) {

        try {
            MyClassVisitor visitor = new MyClassVisitor();

            ClassReader classReader = new ClassReader(in);
            classReader.accept(visitor, ClassReader.SKIP_DEBUG);

            return visitor.getJavaClass();

        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private static JavaVisibility parseVisibility(int access) {
        if ((access & Opcodes.ACC_PUBLIC) != 0)
            return JavaVisibility.PUBLIC;
        else if ((access & Opcodes.ACC_PROTECTED) != 0)
            return JavaVisibility.PROTECTED;
        else if ((access & Opcodes.ACC_PRIVATE) != 0)
            return JavaVisibility.PRIVATE;
        else return JavaVisibility.PACKAGE;
    }

    private static final class MyClassVisitor extends ClassVisitor {

        private JavaClass javaClass;

        public MyClassVisitor() {
            super(Opcodes.ASM5);
        }

        @Override
        public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
            if (javaClass != null)
                throw new IllegalStateException("classMetadata already initialized");

            // TODO: use access
            this.javaClass = new JavaClass(
                    TypeParser.parseObjectType(name).toBasicType().toString(),
                    TypeParser.parseObjectType(superName).toBasicType(),
                    Stream.of(interfaces).map(TypeParser::parseObjectType).collect(toList()));
        }

        @Override
        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
            JavaAnnotation annotation = new JavaAnnotation(TypeParser.parseBasicTypeDescriptor(desc));
            getJavaClass().addAnnotation(annotation);

            return new MyAnnotationVisitor(annotation);
        }

        @Override
        public FieldVisitor visitField(int access, String name, String desc, String signature, Object value) {
            JavaField field = new JavaField(name, parseVisibility(access), TypeParser.parseJavaType(desc, signature), access);
            getJavaClass().addField(field);
            return new MyFieldVisitor(field);
        }

        @Override
        public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
            MethodSignature methodSignature = TypeParser.parseMethodSignature(desc, signature);

            JavaMethod method = new JavaMethod(name, parseVisibility(access), methodSignature.getReturnType(), methodSignature.getParameters(), access);
            getJavaClass().addMethod(method);
            return new MyMethodVisitor(method);
        }

        public JavaClass getJavaClass() {
            if (javaClass == null)
                throw new IllegalStateException("no ClassMetadata available");

            return javaClass;
        }
    }

    private static final class MyAnnotationVisitor extends AnnotationVisitor {

        private final JavaAnnotation annotation;

        public MyAnnotationVisitor(JavaAnnotation annotation) {
            super(Opcodes.ASM5);

            this.annotation = requireNonNull(annotation);
        }

        @Override
        public void visit(String name, Object value) {
            annotation.setAttribute(name, value);
        }

        @Override
        public void visitEnum(String name, String desc, String value) {
            JavaBasicType enumType = TypeParser.parseBasicTypeDescriptor(desc);

            annotation.setAttribute(name, new EnumValue(enumType, value));
        }

        @Override
        public AnnotationVisitor visitAnnotation(String name, String desc) {
            JavaAnnotation nested = new JavaAnnotation(TypeParser.parseTypeDescriptor(desc).toBasicType());
            annotation.setAttribute(name, nested);
            return new MyAnnotationVisitor(nested);
        }

        @Override
        public AnnotationVisitor visitArray(String name) {
            return new ArrayAnnotationVisitor(annotation, name);
        }
    }

    private static final class ArrayAnnotationVisitor extends AnnotationVisitor {
        private final JavaAnnotation annotation;
        private final String name;
        private final List<Object> values = new ArrayList<>();

        public ArrayAnnotationVisitor(JavaAnnotation annotation, String name) {
            super(Opcodes.ASM5);
            this.name = requireNonNull(name);
            this.annotation = requireNonNull(annotation);
        }

        @Override
        public void visitEnd() {
            annotation.setAttribute(name, values.toArray());
        }

        @Override
        public void visit(String name, Object value) {
            values.add(value);
        }

        @Override
        public void visitEnum(String name, String desc, String value) {
            JavaBasicType enumType = TypeParser.parseBasicTypeDescriptor(desc);

            values.add(new EnumValue(enumType, value));
        }

        @Override
        public AnnotationVisitor visitAnnotation(String name, String desc) {
            JavaAnnotation nested = new JavaAnnotation(TypeParser.parseBasicTypeDescriptor(desc));
            values.add(nested);
            return new MyAnnotationVisitor(nested);
        }
    }

    private static final class MyFieldVisitor extends FieldVisitor {

        private final JavaField field;

        public MyFieldVisitor(JavaField field) {
            super(Opcodes.ASM5);
            this.field = field;
        }

        @Override
        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
            JavaAnnotation annotation = new JavaAnnotation(TypeParser.parseBasicTypeDescriptor(desc));
            field.addAnnotation(annotation);
            return new MyAnnotationVisitor(annotation);
        }
    }

    private static final class MyMethodVisitor extends MethodVisitor {

        private final JavaMethod method;
        private int parameterIndex = 0;

        public MyMethodVisitor(JavaMethod method) {
            super(Opcodes.ASM5);
            this.method = method;
        }

        @Override
        public AnnotationVisitor visitAnnotation(String desc, boolean visible) {
            JavaAnnotation annotation = new JavaAnnotation(TypeParser.parseBasicTypeDescriptor(desc));
            method.addAnnotation(annotation);
            return new MyAnnotationVisitor(annotation);
        }

        @Override
        public void visitParameter(String name, int access) {
            JavaParameter nextParameter = method.getParameters().get(parameterIndex++);

            if (name != null)
                nextParameter.initName(name);
        }

        @Override
        public AnnotationVisitor visitParameterAnnotation(int parameter, String desc, boolean visible) {
            JavaParameter javaParameter = method.getParameters().get(parameter);

            JavaAnnotation annotation = new JavaAnnotation(TypeParser.parseBasicTypeDescriptor(desc));
            javaParameter.addAnnotation(annotation);

            return new MyAnnotationVisitor(annotation);
        }
    }
}

File: apina-core/src/main/kotlin/fi/evident/apina/java/reader/ClassMetadataReader.kt
package fi.evident.apina.java.reader

import fi.evident.apina.java.model.*
import fi.evident.apina.java.model.type.JavaType
import fi.evident.apina.java.model.type.TypeSchema
import org.objectweb.asm.*
import java.io.InputStream
import java.util.*

/**
 * Reads class metadata.
 */
internal object ClassMetadataReader {

    fun loadMetadata(inputStream: InputStream): JavaClass {
        val visitor = MyClassVisitor()

        val classReader = ClassReader(inputStream)
        classReader.accept(visitor, ClassReader.SKIP_FRAMES)

        return visitor.getJavaClass()
    }

    private class MyClassVisitor : ClassVisitor(Opcodes.ASM9) {

        private var javaClass: JavaClass? = null

        override fun visit(version: Int, access: Int, name: String, signature: String?, superName: String?, interfaces: Array<String>) {
            check(javaClass == null) { "classMetadata already initialized" }

            var type = parseObjectType(name)
            val superType: JavaType
            val interfaceTypes: List<JavaType>
            val schema: TypeSchema

            if (signature != null) {
                // If we have a generic signature available, parse it using our visitor
                val visitor = ClassSignatureVisitor.parse(signature)

                schema = visitor.schema
                if (!schema.isEmpty)
                    type = JavaType.Parameterized(type, schema.variables)

                superType = visitor.superClass ?: JavaType.Basic(Any::class.java)
                interfaceTypes = visitor.interfaces


            } else {
                superType = superName?.let(::parseObjectType) ?: JavaType.Basic("java.lang.Object")
                interfaceTypes = interfaces.map(::parseObjectType)
                schema = TypeSchema()
            }

            this.javaClass = JavaClass(type, superType, interfaceTypes, access, schema)
        }

        override fun visitAnnotation(desc: String, visible: Boolean): AnnotationVisitor {
            val annotation = JavaAnnotation(parseBasicTypeDescriptor(desc))
            getJavaClass().addAnnotation(annotation)

            return MyAnnotationVisitor(annotation)
        }

        override fun visitField(access: Int, name: String, desc: String, signature: String?, value: Any?): FieldVisitor {
            val field = JavaField(name, JavaVisibility.fromAccessFlags(access), parseJavaType(desc, signature), access)
            getJavaClass().addField(field)
            return MyFieldVisitor(field)
        }

        override fun visitMethod(access: Int, name: String, desc: String, signature: String?, exceptions: Array<String>?): MethodVisitor? {
            val javaClass = getJavaClass()
            if (javaClass.isEnum && name == "<init>") {
                // Skip constructors of enums
                return null
            }

            val methodSignature = parseMethodSignature(desc, signature)

            val method = JavaMethod(javaClass, name, JavaVisibility.fromAccessFlags(access), methodSignature.returnType, methodSignature.parameters, access, methodSignature.schema)
            javaClass.addMethod(method)
            return MyMethodVisitor(method)
        }

        fun getJavaClass(): JavaClass =
            javaClass ?: error("no ClassMetadata available")
    }

    private class MyAnnotationVisitor(private val annotation: JavaAnnotation) : AnnotationVisitor(Opcodes.ASM9) {

        override fun visit(name: String, value: Any) {
            annotation.setAttribute(name, if (value is Type) javaType(value) else value)
        }

        override fun visitEnum(name: String, desc: String, value: String) {
            val enumType = parseBasicTypeDescriptor(desc)

            annotation.setAttribute(name, EnumValue(enumType, value))
        }

        override fun visitAnnotation(name: String, desc: String): AnnotationVisitor {
            val nested = JavaAnnotation(parseTypeDescriptor(desc).toBasicType())
            annotation.setAttribute(name, nested)
            return MyAnnotationVisitor(nested)
        }

        override fun visitArray(name: String): AnnotationVisitor {
            return ArrayAnnotationVisitor(annotation, name)
        }
    }

    private class ArrayAnnotationVisitor(private val annotation: JavaAnnotation, private val name: String) : AnnotationVisitor(Opcodes.ASM9) {
        private val values = ArrayList<Any>()

        override fun visitEnd() {
            annotation.setAttribute(name, values.toTypedArray())
        }

        override fun visit(name: String?, value: Any) {
            values.add(value)
        }

        override fun visitEnum(name: String?, desc: String, value: String) {
            val enumType = parseBasicTypeDescriptor(desc)

            values.add(EnumValue(enumType, value))
        }

        override fun visitAnnotation(name: String?, desc: String): AnnotationVisitor {
            val nested = JavaAnnotation(parseBasicTypeDescriptor(desc))
            values.add(nested)
            return MyAnnotationVisitor(nested)
        }
    }

    private class MyFieldVisitor(private val field: JavaField) : FieldVisitor(Opcodes.ASM9) {

        override fun visitAnnotation(desc: String, visible: Boolean): AnnotationVisitor {
            val annotation = JavaAnnotation(parseBasicTypeDescriptor(desc))
            field.addAnnotation(annotation)
            return MyAnnotationVisitor(annotation)
        }
    }

    private class MyMethodVisitor(private val method: JavaMethod) : MethodVisitor(Opcodes.ASM9) {
        private var parameterIndex = 0

        /** For each parameter the table contains the index of local variable describing the parameter  */
        private val parameterIndicesInLVT: IntArray

        init {
            this.parameterIndicesInLVT = initLVTParameterIndices(method.isStatic, method.parameters)
        }

        override fun visitAnnotation(desc: String, visible: Boolean): AnnotationVisitor {
            val annotation = JavaAnnotation(parseBasicTypeDescriptor(desc))
            method.addAnnotation(annotation)
            return MyAnnotationVisitor(annotation)
        }

        override fun visitParameter(name: String?, access: Int) {
            val nextParameter = method.parameters[parameterIndex++]

            if (name != null)
                nextParameter.name = name
        }

        override fun visitLocalVariable(name: String, desc: String, signature: String?, start: Label, end: Label, index: Int) {
            for ((i, paramIndex) in parameterIndicesInLVT.withIndex())
                if (paramIndex == index)
                    method.parameters[i].name = name
        }

        override fun visitParameterAnnotation(parameter: Int, desc: String, visible: Boolean): AnnotationVisitor {
            val javaParameter = method.parameters[parameter]

            val annotation = JavaAnnotation(parseBasicTypeDescriptor(desc))
            javaParameter.addAnnotation(annotation)

            return MyAnnotationVisitor(annotation)
        }

        private fun initLVTParameterIndices(isStatic: Boolean, parameters: List<JavaParameter>): IntArray {
            val result = IntArray(parameters.size)

            var index = if (isStatic) 0 else 1
            for (i in result.indices) {
                result[i] = index
                if (parameters[i].type.isWide)
                    index += 2
                else
                    index++
            }
            return result
        }
    }
}
